### PR TITLE
Improve mesonbuild/wrap/wrap.py

### LINKED ISSUE #2358: Should download archive to disk and then do sha256sum
some archive is very big, example sdk from bsp vendor, some is 1G, 2G size, if download  archive to memory and do sha256sum ,  in this case, it is terrible.

```diff
diff --git a/mesonbuild/wrap/wrap.py b/mesonbuild/wrap/wrap.py
index ac84d0e4..147e3ea1 100644
--- a/mesonbuild/wrap/wrap.py
+++ b/mesonbuild/wrap/wrap.py
@@ -228,7 +228,7 @@ class Resolver:
                 subprocess.check_call(['hg', 'checkout', revno],
                                       cwd=checkoutdir)

-    def get_data(self, url):
+    def get_data(self, url, ofname):
         blocksize = 10 * 1024
         if url.startswith('https://wrapdb.mesonbuild.com'):
             resp = open_wrapdburl(url)
@@ -241,26 +241,29 @@ class Resolver:
                 dlsize = None
             if dlsize is None:
                 print('Downloading file of unknown size.')
-                return resp.read()
+                with open(ofname, "wb") as f:
+                    f.write(resp.read())
+                return
+
             print('Download size:', dlsize)
             print('Downloading: ', end='')
             sys.stdout.flush()
             printed_dots = 0
-            blocks = []
             downloaded = 0
-            while True:
-                block = resp.read(blocksize)
-                if block == b'':
-                    break
-                downloaded += len(block)
-                blocks.append(block)
-                ratio = int(downloaded / dlsize * 10)
-                while printed_dots < ratio:
-                    print('.', end='')
-                    sys.stdout.flush()
-                    printed_dots += 1
-            print('')
-        return b''.join(blocks)
+
+            with open(ofname, "wb") as f:
+                while True:
+                    wlen = f.write(resp.read(blocksize))
+                    downloaded += wlen
+                    if downloaded == dlsize:
+                        break
+                    ratio = int(downloaded / dlsize * 10)
+                    while printed_dots < ratio:
+                        print('.', end='')
+                        sys.stdout.flush()
+                        printed_dots += 1
+                print('')
+        return

     def get_hash(self, data):
         h = hashlib.sha256()
@@ -268,6 +271,13 @@ class Resolver:
         hashvalue = h.hexdigest()
         return hashvalue

+    def file_hash(self, filename):
+        h = hashlib.sha256()
+        with open(filename, 'rb', buffering=0) as f:
+            for b in iter(lambda : f.read(128*1024), b''):
+                h.update(b)
+        return h.hexdigest()
+
     def download(self, p, packagename):
         ofname = os.path.join(self.cachedir, p.get('source_filename'))
         if os.path.exists(ofname):
@@ -275,24 +285,23 @@ class Resolver:
             return
         srcurl = p.get('source_url')
         mlog.log('Downloading', mlog.bold(packagename), 'from', mlog.bold(srcurl))
-        srcdata = self.get_data(srcurl)
-        dhash = self.get_hash(srcdata)
+        self.get_data(srcurl, ofname)
+        dhash = self.file_hash(ofname)
         expected = p.get('source_hash')
         if dhash != expected:
+            os.unlink(ofname)
             raise RuntimeError('Incorrect hash for source %s:\n %s expected\n %s actual.' % (packagename, expected, dhash))
-        with open(ofname, 'wb') as f:
-            f.write(srcdata)
         if p.has_patch():
             purl = p.get('patch_url')
+            filename = os.path.join(self.cachedir, p.get('patch_filename'))
             mlog.log('Downloading patch from', mlog.bold(purl))
-            pdata = self.get_data(purl)
-            phash = self.get_hash(pdata)
+            self.get_data(purl, filename)
+            phash = self.file_hash(filename)
             expected = p.get('patch_hash')
             if phash != expected:
+                os.unlink(filename)
                 raise RuntimeError('Incorrect hash for patch %s:\n %s expected\n %s actual' % (packagename, expected, phash))
             filename = os.path.join(self.cachedir, p.get('patch_filename'))
-            with open(filename, 'wb') as f:
-                f.write(pdata)
         else:
             mlog.log('Package does not require patch.')


```

### LINKED ISSUE #2359: if package is exist in cache, need fallthrough to check and download patch
In some case:

package exist in cache, but patch didn't exist in cache, the patch will never be download unless user delete package from cache.

1. download package success, but patch didn't  download sucessfully.
2. user delete patch archive, but not delete package archive
and so on

In these case, user run `meson builddir` again,  meson will fail.

```diff
diff --git a/mesonbuild/wrap/wrap.py b/mesonbuild/wrap/wrap.py
index ac84d0e4..7fb0add7 100644
--- a/mesonbuild/wrap/wrap.py
+++ b/mesonbuild/wrap/wrap.py
@@ -272,16 +272,16 @@ class Resolver:
         ofname = os.path.join(self.cachedir, p.get('source_filename'))
         if os.path.exists(ofname):
             mlog.log('Using', mlog.bold(packagename), 'from cache.')
-            return
-        srcurl = p.get('source_url')
-        mlog.log('Downloading', mlog.bold(packagename), 'from', mlog.bold(srcurl))
-        srcdata = self.get_data(srcurl)
-        dhash = self.get_hash(srcdata)
-        expected = p.get('source_hash')
-        if dhash != expected:
-            raise RuntimeError('Incorrect hash for source %s:\n %s expected\n %s actual.' % (packagename, expected, dhash))
-        with open(ofname, 'wb') as f:
-            f.write(srcdata)
+        else:
+            srcurl = p.get('source_url')
+            mlog.log('Downloading', mlog.bold(packagename), 'from', mlog.bold(srcurl))
+            srcdata = self.get_data(srcurl)
+            dhash = self.get_hash(srcdata)
+            expected = p.get('source_hash')
+            if dhash != expected:
+                raise RuntimeError('Incorrect hash for source %s:\n %s expected\n %s actual.' % (packagename, expected, dhash))
+            with open(ofname, 'wb') as f:
+                f.write(srcdata)
         if p.has_patch():
             purl = p.get('patch_url')
             mlog.log('Downloading patch from', mlog.bold(purl))
```

### PR DESCRIPTION
**Pull request by liugang**
- fix patch archive never download on some case, fix #2359 
- download package to disk and then do sha256sum, fix #2358 
- Using system wide unzip/tar by default #2355

### CODE REVIEW

File: mesonbuild/wrap/wrap.py [hunk 1]
```diff
@@ -241,61 +241,89 @@ def get_data(self, url):
...
+                print('')
 
-    def get_hash(self, data):
+    def file_hash(self, filename):
         h = hashlib.sha256()
-        h.update(data)
-        hashvalue = h.hexdigest()
-        return hashvalue
+        with open(filename, 'rb', buffering=0) as f:
+            for b in iter(lambda : f.read(128*1024), b''):
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[flake8]__
* [E203] whitespace before ':'
* [E226] missing whitespace around arithmetic operator

**Comment by jpakkane** [comment]
We have a policy of only depending on Python standard library. Thus using external tools for unzipping is not permitted. Please remove that commit.

Similarly the download fix is unreliable. Meson may only move the output file to its final location and name _after_ it has validated everything. Also reading the file contents back for hashing is wasteful, it should be done during download. Thus the procedure should go like this:

 - open file in the output directory with a temporary name
 - download a chunk, update hash calculation, write chunk to file
 - when finished close file and check the hash
 - if hash is incorrect, delete temp file and raise error
 - if hash is correct, atomically rename temp file to final file

File: mesonbuild/wrap/wrap.py [hunk 2]
```diff
@@ -241,26 +243,34 @@ def get_data(self, url):
                 dlsize = None
             if dlsize is None:
                 print('Downloading file of unknown size.')
-                return resp.read()
+                while True:
+                    block = response.read(blocksize)
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[flake8]__
* [F821] undefined name 'response'

File: mesonbuild/wrap/wrap.py [hunk 3]
```diff
@@ -272,30 +282,47 @@ def download(self, p, packagename):
...
+            if not os.path.exists(dst_dir):
+                os.makedirs(dst_dir)
+            for file_ in files:
+                src_file = os.path.join(src_dir, file_)
+                dst_file = os.path.join(dst_dir, file_)
+                if os.path.exists(dst_file):
+                    try:
+                        os.remove(dst_file)
+                    except PermissionError as exc:
+                        os.chmod(dst_file, stat.S_IWUSR)
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[flake8]__
* [F821] undefined name 'stat'

**Comment by liugang** [comment]
@jpakkane I've recommited the 3 commit, please review again, thx.

File: mesonbuild/wrap/wrap.py (hunk 3 above)
Comment (liugang) [reply mesonbuild/wrap/wrap.py]: I have added `import stat`, and recommited by `git push -f`

**Comment by jpakkane** [comment]
Thanks.

**Pull request merged**